Move lowerings to `_d_array{setassign,assign_{l,r}}` to a `LoweredAssignExp` AST node

teodutu opened the pull request:
This PR does the following:
- uses a `LoweredAssignExp` node for assignments that lower to `_d_arrayassign_{l,r}`
- removes the code in `dinterpret.d` that recreates the original `AssigExp`
- simplifies `isArrayConstructionOrAssign()` to just `isArrayConstruction()` since assignments are now handled by a separate AST node
- fixes issue 23959

teodutu committed:
Fix Issue 23959 - Move lowerings to `_d_array{setassign,assign_{l,r}}` to a `LoweredAssignExp` AST node

The compiler now uses a `LoweredAssignExp` node for assignmets that
lower to `_d_arrayassign_{l,r}`

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

thewilsonator commented:
Are the things that this PR does separable? If so please separate them out, at least into separate commits.

teodutu committed:
Simplify CTFE handling of `LoweredAssignExp`

`LoweredAssignExp` can be interpreted as regular `AssignExp` without
having to recreate them from the lowering. Therefore, this commit does
the following:
- Remove the code in `dinterpret.d` that recreates the original
`AssigExp`
- Simplify `isArrayConstructionOrAssign()` to just
  `isArrayConstruction()` since assignments are now handled by a separate
AST node

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu commented:
Done @thewilsonator. I moved the changes to CTFE to another commit. I think moving them to another PR would be overkill as using `LoweredAssignExp` implies simplifying the interpretation as well.

CyberShadow commented:
Thank you for the quick fix! I've confirmed that it fixes my real project.

I may be behind the times on processes, but shouldn't regression fixes target `stable`?

teodutu commented:
> I may be behind the times on processes, but shouldn't regression fixes target `stable`?

I think you're right. I've changed the destination branch.

teodutu commented:
@RazvanN7 @dkorpel @thewilsonator please take a look at this when you can.

RazvanN7 commented:
@teodutu please have the commit message contain "Fix Issue 23959" (not "fixes ...") so that the bot automatically closes it when this gets merged.

RazvanN7 commented:
Yes, regressions should target stable, provided that the bug manifests in stable.

teodutu commented:
Done @RazvanN7, but the bot's message still doesn't mark the bug as auto-close. Should I reword the commit message differently?

teodutu commented:
Also, it seems something is wrong with the libc version that's present on the systems that run the buildkite tests [1]:
```
/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found
```
https://buildkite.com/dlang/dmd/builds/31700

CyberShadow commented:
> Done @RazvanN7, but the bot's message still doesn't mark the bug as auto-close. Should I reword the commit message differently?

It has to be at the start of the line: https://github.com/dlang/dlang-bot/blob/9b90a61c94216043159f94d73640455fc46ed9dd/source/dlangbot/bugzilla.d#L34

So I guess the problem is that it's in a Markdown list.

> `` /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found ``

Hm, that error means that a it tried to run a program (dmd?) which was compiled against a newer libc than the one on the current system. Usually this is why things which link to glibc are built on a system that's as old as possible. Maybe someone updated the system on which dmd is built?

teodutu commented:
Done @CyberShadow, now the auto-close problem is solved. Thanks!

RazvanN7 commented:
@teodutu rebasing will fix the existing buildkite issues.

teodutu commented:
Done @RazvanN7

RazvanN7 commented on compiler/src/dmd/dinterpret.d:
```diff
@@ -3891,7 +3891,7 @@ public:
             assignInPlace(oldval, newval);
         }
-        else if (wantCopy && e.op == EXP.assign)
+        else if (wantCopy && (e.op == EXP.assign || e.op == EXP.loweredAssignExp))
```
Maybe I am missing something here, but as far as I understand the interpreter should never have to deal with `LoweredAssignExp`.  That's why we are keeping the original expression, no?

teodutu replied:
No, it should. In `expressionsem.d` the lowering changes the expression to `LoweredAssignExp` [1]. That's why the interpreter should handle it the same way it handles `AssignExp`s.

[1] https://github.com/dlang/dmd/blob/a528c995bf1c80acd4766002d9da198ab396989c/compiler/src/dmd/expressionsem.d#L10311-L10312

RazvanN7 commented on compiler/src/dmd/dinterpret.d:
```diff
@@ -4792,17 +4791,14 @@ public:
 
                 Expression eb = (*e.arguments)[1];
-                if (eb.isCastExp() && fd.ident != Id._d_arraysetctor)
+                if (eb.isCastExp && fd.ident == Id._d_arrayctor)
```
```suggestion
                if (eb.isCastExp() && fd.ident == Id._d_arrayctor)
```
